proxy/pkg/zdmproxy/streamidmapper.go

```diff
@@ -0,0 +1,130 @@
+package zdmproxy
+
+import (
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"sync"
+)
+
+var maxStreamIds int16 = 2048
+
+// StreamIdMapper is used to map the incoming stream ids from the client/driver to internal ids managed by the proxy
+// This is required because we also generate requests inside the proxy that go to the cluster through the same connection,
+// hence they must have non-conflicting ids with user's requests.
+type StreamIdMapper interface {
+	GetNewIdFor(streamId int16) (int16, error)
+	RestoreId(syntheticId int16) (int16, error)
+	ReleaseId(syntheticId int16) (int16, error)
+}
+
+type cqlStreamIdMapper struct {
+	sync.Mutex
+	ids chan int16
+}
+
+func NewCqlStreamIdMapper(maxStreamIds int) StreamIdMapper {
+	var ids = make(chan int16, maxStreamIds)
+	for i := int16(0); i < int16(maxStreamIds); i++ {
+		ids <- i
+	}
+	return &cqlStreamIdMapper{
+		ids: ids,
+	}
+}
+
+func (icsim *cqlStreamIdMapper) GetNewIdFor(_ int16) (int16, error) {
+	icsim.Lock()
+	defer icsim.Unlock()
+	select {
+	case id, ok := <-icsim.ids:
+		if ok {
+			return id, nil
+		} else {
+			return -1, fmt.Errorf("stream id channel closed")
+		}
+	default:
+		return -1, fmt.Errorf("no stream id available")
+	}
+}
+
+func (icsim *cqlStreamIdMapper) RestoreId(_ int16) (int16, error) {
+	return -1, fmt.Errorf("internal CQL ids have no original ids to be restored")
+}
+
+func (icsim *cqlStreamIdMapper) ReleaseId(syntheticId int16) (int16, error) {
+	icsim.Lock()
+	defer icsim.Unlock()
+	select {
+	case icsim.ids <- syntheticId:
+	default:
+		log.Errorf("stream ids channel full, ignoring id %v", syntheticId)
+	}
+	return -1, nil
+}
+
+type streamIdMapper struct {
+	sync.Mutex
+	idMapper   map[int16]int16
+	synMapper  map[int16]int16
+	clusterIds chan int16
+}
+
+func NewStreamIdMapper(maxStreamIds int) StreamIdMapper {
+	idMapper := make(map[int16]int16)
+	synMapper := make(map[int16]int16)
+	streamIdsQueue := make(chan int16, maxStreamIds)
+	for i := int16(0); i < int16(maxStreamIds); i++ {
+		streamIdsQueue <- i
+	}
+	return &streamIdMapper{
+		idMapper:   idMapper,
+		clusterIds: streamIdsQueue,
+		synMapper:  synMapper,
+	}
+}
+
+func (sim *streamIdMapper) GetNewIdFor(streamId int16) (int16, error) {
+	sim.Lock()
+	defer sim.Unlock()
```

Comment:
We should try to reduce locking in this, I think we can use the `sync.Map` instead of regular map and instead of removing entries from the map we can set them to nil or an empty global object (this is because `sync.Map` does not perform well with deletes).

An alternative is to just move the locking to the part where the maps are being updated (possibly remove the `contains` check altogether - check my comment below) and benchmark to see if it's good enough.

Reply:
Update: I think we can just move the locking to the actual map operation part (after reading from the channel) and then we can see how it performs before thinking about optimizations

Reply:
> Update: I think we can just move the locking to the actual map operation part (after reading from the channel) and then we can see how it performs before thinking about optimizations

Done